### PR TITLE
 JIRA:1060 - Modified KNN to accept expressions in point_column_name …

### PR DESCRIPTION
**Pull request by hpandeycodeit**
JIRA: MADLIB-1060

This PR contains changes for KNN: 

1. Modified code - knn.py_in,  to support regular PostgreSQL expressions for point_column_name and test_column_name parameters. 
2. Added test cases for the above scenarios.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/659/

File: src/ports/postgres/modules/knn/knn.py_in [hunk 1]
```diff
@@ -53,22 +55,10 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
...
-    if not array_col_has_no_null(point_source, point_column_name):
-        plpy.error("kNN Error: Feature column '{0}' in train table has some"
-                   " NULL values.".format(point_column_name))
-    if not array_col_has_no_null(test_source, test_column_name):
-        plpy.error("kNN Error: Feature column '{0}' in test table has some"
-                   " NULL values.".format(test_column_name))
+
+    _assert(point_column_name, "KNN error: Invalid point_column_name expression")
+
+    _assert(test_column_name, "KNN error: Invalid test_column_name expression")
```
Comment (njayaram2) [inline src/ports/postgres/modules/knn/knn.py_in]: `KNN` in the error message is different from `kNN` used in other error messages (capital `K`). Please keep it consistent as `kNN`.

File: src/ports/postgres/modules/knn/test/knn.sql_in [hunk 2]
```diff
@@ -122,5 +122,13 @@ select knn('knn_train_data_reg','data','id','label','knn_test_data','data','id',
...
 
 
+drop table if exists madlib_knn_result_classification;
+select knn('knn_train_data','data[1:1]||data[2:2]','id','label','knn_test_data','data[1:1]||data[2:2]','id','madlib_knn_result_classification',3,False,'MADLIB_SCHEMA.squared_dist_norm2', True);
+select assert(array_agg(prediction::numeric order by id)='{1,1,0,1,0,0}', 'Wrong output in classification') from madlib_knn_result_classification;
+
+drop table if exists madlib_knn_result_regression;
+select knn('knn_train_data_reg','data[1:1]||data[2:2]','id','label','knn_test_data','data[1:1]||data[2:2]','id','madlib_knn_result_regression',3,False,'MADLIB_SCHEMA.squared_dist_norm2', True);
+select assert(array_agg(prediction::numeric order by id)='{1,1,0.0408728591876018,1,0,0}', 'Wrong output in regression') from madlib_knn_result_regression;
+
```
Comment (njayaram2) [inline src/ports/postgres/modules/knn/test/knn.sql_in]: We could add another test case, where we use `ARRAY(d1, d2)` as the expression, say for `point_column_name`. Of course, you will then have to add two new columns d1 and d2 while creating the input table `knn_train_data` in that case.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 3]
```diff
@@ -264,12 +275,14 @@ def knn(schema_madlib, point_source, point_column_name, point_id,
...
-                                test.{test_column_name})
+                                train.{point_col_name_temp},
+                                test.{test_col_name_temp})
                             AS dist
                             {label_out}
-                            FROM {point_source} AS train,
-                                 {test_source} AS test
+                            FROM
+                                 {point_source_temp_table} as train,
+                                 {test_source_temp_table} as test
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: Replaced the temp tables with subqueries.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 4]
```diff
@@ -290,11 +303,11 @@ def knn(schema_madlib, point_source, point_column_name, point_id,
                 ON knn_temp.{test_id_temp} = knn_test.{test_id}
                     {view_join}
                 GROUP BY knn_temp.{test_id_temp},
-                         knn_test.{test_column_name}
+                    {test_column_name}
                          {view_grp_by}
             """
         plpy.execute(sql.format(**locals()))
-        plpy.execute("DROP TABLE IF EXISTS {0}".format(interim_table))
+        # plpy.execute("DROP TABLE IF EXISTS {0}".format(interim_table))
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: I pushed by mistake. I have reverted the changes.

File: src/ports/postgres/modules/knn/test/knn.sql_in (hunk 2 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/test/knn.sql_in]: Added this test as well along with other minor changes as suggested.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/664/

File: src/ports/postgres/modules/knn/knn.py_in [hunk 5]
```diff
@@ -53,22 +55,12 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
...
-    if not is_col_array(test_source, test_column_name):
-        plpy.error("kNN Error: Feature column '{0}' in test table is not"
-                   " an array.".format(test_column_name))
-
-    if not array_col_has_no_null(point_source, point_column_name):
-        plpy.error("kNN Error: Feature column '{0}' in train table has some"
-                   " NULL values.".format(point_column_name))
-    if not array_col_has_no_null(test_source, test_column_name):
-        plpy.error("kNN Error: Feature column '{0}' in test table has some"
-                   " NULL values.".format(test_column_name))
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: Added it back. Had to do a minor change in array_col_has_no_null() for the expressions.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 6]
```diff
@@ -53,22 +55,12 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
...
         cols_in_tbl_valid(point_source, [label_column_name], 'kNN')
-    cols_in_tbl_valid(point_source, (point_column_name, point_id), 'kNN')
-    cols_in_tbl_valid(test_source, (test_column_name, test_id), 'kNN')
-
-    if not is_col_array(point_source, point_column_name):
-        plpy.error("kNN Error: Feature column '{0}' in train table is not"
-                   " an array.".format(point_column_name))
-    if not is_col_array(test_source, test_column_name):
-        plpy.error("kNN Error: Feature column '{0}' in test table is not"
-                   " an array.".format(test_column_name))
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: As per the Jira and our discussion, it is okay to fail at distance function if the point_column_name and test_column_name params are not arrays.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 7]
```diff
@@ -264,12 +260,17 @@ def knn(schema_madlib, point_source, point_column_name, point_id,
...
                             {label_out}
-                            FROM {point_source} AS train,
-                                 {test_source} AS test
+                            FROM
+                            (
+                            SELECT {point_id} , {point_column_name} as p_col_name , {label_column_name} from {point_source}
+                            ) train,
+                            (
+                            SELECT {test_id} ,{test_column_name} as t_col_name from {test_source}
+                            ) test
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: Done.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 8]
```diff
@@ -53,22 +55,12 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
 
     if label_column_name and label_column_name.strip():
         cols_in_tbl_valid(point_source, [label_column_name], 'kNN')
-    cols_in_tbl_valid(point_source, (point_column_name, point_id), 'kNN')
-    cols_in_tbl_valid(test_source, (test_column_name, test_id), 'kNN')
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/knn/knn.py_in]: Added these functions back.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/670/

File: src/ports/postgres/modules/knn/knn.py_in (hunk 6 above)
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/ports/postgres/modules/knn/knn.py_in]: I was just playing with the function to see how user-friendly the error message would be. Found that, with the older kNN version, if a column was not of array type, the error was informative:
`kNN Error: Feature column 'y' in train table is not an array.`
With the code in this PR, the error for the same failure is:
`function array_upper(double precision, integer) does not exist`

I don't think this is informative enough for users. I would surely like to continue the discussion that has already happened on this. @fmcquillan99 and @hpandeycodeit , any thoughts?

[the remark this answers is not in the document]
Comment (fmcquillan99) [reply src/ports/postgres/modules/knn/knn.py_in]: Yes I agree that this error message is too vague.  So in retrospect I suggest we do check that a valid array type is generated by the expression.   Sorry if I mislead earlier @hpandeycodeit 

@njayaram2 do you have an existing function that we can use for this check?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/671/

**Comment by fmcquillan99** [comment]
load data:
```
DROP TABLE IF EXISTS knn_train_data;

CREATE TABLE knn_train_data (
                    id integer, 
                    data integer[], 
                    label integer  -- Integer label means for classification
                    );

INSERT INTO knn_train_data VALUES
(1, '{1,1}', 1),
(2, '{2,2}', 1),
(3, '{3,3}', 1),
(4, '{4,4}', 1),
(5, '{4,5}', 1),
(6, '{20,50}', 0),
(7, '{10,31}', 0),
(8, '{81,13}', 0),
(9, '{1,111}', 0);
```

run knn to list nearest neighbors only
(without doing classification or regression). 
```
DROP TABLE IF EXISTS knn_result_list_neighbors;

SELECT * FROM madlib.knn(
                'knn_train_data_reg',  -- Table of training data
                'data',                -- Col name of training data
                'id',                  -- Col Name of id in train data
                NULL,                  -- NULL training labels means just list neighbors
                'knn_test_data',       -- Table of test data
                'data',                -- Col name of test data
                'id',                  -- Col name of id in test data
                'knn_result_list_neighbors', -- Output table
                3                      -- Number of nearest neighbors
                );
```
results in error
```
ERROR:  plpy.SPIError: column "none" does not exist
LINE 22: ...b_temp_p_col_name68224549_1536338162_43630832__ , None from ...
                                                              ^
QUERY:  
            CREATE TEMP TABLE __madlib_temp_interim_table80328381_1536338162_13726874__ AS
                SELECT * FROM (
                    SELECT row_number() over
                            (partition by __madlib_temp_test_id_temp32938499_1536338162_42836514__ order by __madlib_temp_dist75319423_1536338162_4342866__) AS __madlib_temp_r77948314_1536338162_302550__,
                            __madlib_temp_test_id_temp32938499_1536338162_42836514__,
                            __madlib_temp_train_id51898468_1536338162_48880352__,
                            CASE WHEN __madlib_temp_dist75319423_1536338162_4342866__ = 0.0 THEN 1000000.0
                                 ELSE 1.0 / __madlib_temp_dist75319423_1536338162_4342866__
                            END AS __madlib_temp_dist_inverse21927322_1536338162_30562226__
                            
                    FROM (
                        SELECT __madlib_temp_test58595915_1536338162_24414359__.id AS __madlib_temp_test_id_temp32938499_1536338162_42836514__,
                            __madlib_temp_train73645570_1536338162_46994036__.id as __madlib_temp_train_id51898468_1536338162_48880352__,
                            madlib.squared_dist_norm2(
                                __madlib_temp_p_col_name68224549_1536338162_43630832__,
                                __madlib_temp_t_col_name86464547_1536338162_93305987__)
                            AS __madlib_temp_dist75319423_1536338162_4342866__
                            
                            FROM
                            (
                            SELECT id , data as __madlib_temp_p_col_name68224549_1536338162_43630832__ , None from knn_train_data_reg
                            ) __madlib_temp_train73645570_1536338162_46994036__,
                            (
                            SELECT id ,data as __madlib_temp_t_col_name86464547_1536338162_93305987__ from knn_test_data
                            ) __madlib_temp_test58595915_1536338162_24414359__
                        ) __madlib_temp_x_temp_table56760627_1536338162_7182722__
                    ) __madlib_temp_y_temp_table96170617_1536338162_37029044__
            WHERE __madlib_temp_y_temp_table96170617_1536338162_37029044__.__madlib_temp_r77948314_1536338162_302550__ <= 3
            
CONTEXT:  Traceback (most recent call last):
  PL/Python function "knn", line 33, in <module>
    weighted_avg
  PL/Python function "knn", line 305, in knn
PL/Python function "knn"
SQL statement "SELECT  madlib.knn( $1 , $2 , $3 , $4 , $5 , $6 , $7 , $8 , $9 ,TRUE, 'madlib.squared_dist_norm2', FALSE)"
PL/pgSQL function "knn" line 4 at assignment
```

**Comment by njayaram2** [comment]
@fmcquillan99 thanks for testing this out. I can have a look at this issue.

**Comment by fmcquillan99** [comment]
I'm not sure what this is doing:
```
%%sql
DROP TABLE IF EXISTS knn_result_classification;

SELECT * FROM madlib.knn(
                'knn_train_data',      -- Table of training data
                'array[99.]::int[] || array[99]',                -- Col name of training data
                'id',                  -- Col name of id in train data
                'label',               -- Training labels
                'knn_test_data',       -- Table of test data
                'data',                -- Col name of test data
                'id',                  -- Col name of id in test data
                'knn_result_classification',  -- Output table
                 1,                    -- Number of nearest neighbors
                 True,                 -- True to list nearest-neighbors by id
                 'madlib.squared_dist_norm2' -- Distance function
                );

SELECT * from knn_result_classification ORDER BY id;
``` 
produces
```
 id |  data   | prediction | k_nearest_neighbours 
----+---------+------------+----------------------
  1 | {2,1}   |          0 | {8}
  2 | {2,6}   |          0 | {8}
  3 | {15,40} |          0 | {8}
  4 | {12,1}  |          0 | {8}
  5 | {2,90}  |          1 | {1}
  6 | {50,45} |          1 | {1}
(6 rows)
```

I get the same result if I do:
```
DROP TABLE IF EXISTS knn_result_classification;

SELECT * FROM madlib.knn(
                'knn_train_data',      -- Table of training data
                'array[0.]::int[] || array[0]',                -- Col name of training data
                'id',                  -- Col name of id in train data
                'label',               -- Training labels
                'knn_test_data',       -- Table of test data
                'data',                -- Col name of test data
                'id',                  -- Col name of id in test data
                'knn_result_classification',  -- Output table
                 1,                    -- Number of nearest neighbors
                 True,                 -- True to list nearest-neighbors by id
                 'madlib.squared_dist_norm2' -- Distance function
                );

SELECT * from knn_result_classification ORDER BY id;
```
gives
```
 id |  data   | prediction | k_nearest_neighbours 
----+---------+------------+----------------------
  1 | {2,1}   |          0 | {8}
  2 | {2,6}   |          0 | {8}
  3 | {15,40} |          0 | {8}
  4 | {12,1}  |          0 | {8}
  5 | {2,90}  |          1 | {1}
  6 | {50,45} |          1 | {1}
(6 rows)
```

**Comment by fmcquillan99** [comment]
(1)
expression for test data array:
```
DROP TABLE IF EXISTS knn_result_classification;

SELECT * FROM madlib.knn(
                'knn_train_data',      -- Table of training data
                'data',                -- Col name of training data
                'id',                  -- Col name of id in train data
                'label',               -- Training labels
                'knn_test_data',       -- Table of test data
                '3 || ARRAY[4]',                -- Col name of test data
                'id',                  -- Col name of id in test data
                'knn_result_classification',  -- Output table
                 3,                    -- Number of nearest neighbors
                 True,                 -- True to list nearest-neighbors by id
                 'madlib.squared_dist_norm2' -- Distance function
                );

SELECT * from knn_result_classification ORDER BY id;
```
produces
```
 id | 3 || ARRAY[4] | prediction | k_nearest_neighbours 
----+---------------+------------+----------------------
  1 | {3,4}         |          1 | {3,4,5}
  2 | {3,4}         |          1 | {3,4,5}
  3 | {3,4}         |          1 | {3,4,5}
  4 | {3,4}         |          1 | {4,3,5}
  5 | {3,4}         |          1 | {3,4,5}
  6 | {3,4}         |          1 | {4,3,5}
(6 rows)
```


(2)
another expression for test data array:
```
DROP TABLE IF EXISTS knn_result_classification;

SELECT * FROM madlib.knn(
                'knn_train_data',      -- Table of training data
                'data',                -- Col name of training data
                'id',                  -- Col name of id in train data
                'label',               -- Training labels
                'knn_test_data',       -- Table of test data
                'array[3.]::int[] || array[4]',                -- Col name of test data
                'id',                  -- Col name of id in test data
                'knn_result_classification',  -- Output table
                 3,                    -- Number of nearest neighbors
                 True,                 -- True to list nearest-neighbors by id
                 'madlib.squared_dist_norm2' -- Distance function
                );

SELECT * from knn_result_classification ORDER BY id;
```
produces
```
 id | array[3.]::int[] || array[4] | prediction | k_nearest_neighbours 
----+------------------------------+------------+----------------------
  1 | {3,4}                        |          1 | {3,4,5}
  2 | {3,4}                        |          1 | {3,4,5}
  3 | {3,4}                        |          1 | {4,3,5}
  4 | {3,4}                        |          1 | {3,4,5}
  5 | {3,4}                        |          1 | {4,3,5}
  6 | {3,4}                        |          1 | {4,3,5}
(6 rows)
```
so this bit seems to work

**Comment by fmcquillan99** [comment]
Actually the earlier issue above ^^^ is OK, where I said `I'm not sure what this is doing` because forcing all training data to be a single point means that the distance to all test points is identical.  So a random nearest neighbor could be picked.  Which it is what seems to be happening.  

So I think just fix the error above and this should be good to go.

**Comment by hpandeycodeit** [comment]
Thanks for the update @fmcquillan99! I have fixed the error above and added a test case as well.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/672/

**Comment by fmcquillan99** [comment]
re-running the failed test, seems to pass now:

```
SELECT * FROM knn_result_list_neighbors ORDER BY id;
```
produces
```
 id |  data   | k_nearest_neighbours 
----+---------+----------------------
  1 | {2,1}   | {1,2,3}
  2 | {2,6}   | {5,4,3}
  3 | {15,40} | {7,6,5}
  4 | {12,1}  | {4,5,3}
  5 | {2,90}  | {9,6,7}
  6 | {50,45} | {6,7,8}
(6 rows)
```

LGTM

**Pull request closed without merging**